pydra/engine/specs.py

```diff
@@ -50,6 +50,8 @@ def hash(self):
                 continue
             if field.type == File and "container_path" not in field.metadata:
                 inp_dict[field.name] = hash_file(getattr(self, field.name))
+            elif field.type == ty.List[File] or field.type == ty.Tuple[File]:
```

Comment:
what if it was a `List[List[File]`?

it's likely that this bit has to be refactored into a recursive function call. also should support dictionaries whose values are `File`.  this is the function from nipype:

https://github.com/nipy/nipype/blob/master/nipype/interfaces/base/specs.py#L273

i don't think some of the elements are necessary with py3, but the general idea is still something to consider here.